server/services/clean-up.js

```diff
@@ -57,9 +57,130 @@ const cleanStaleInProgressParticipant = async () => {
+      WHERE p.id = ps.participant_id
+    );
+  `;
+  await dbClient.db.query(pushUnengagedParticipants);
+
+  const pushRejectedParticipants = `
+    INSERT INTO expired_participants_table
+    SELECT
+      p.id, max(ps.created_at) AS last_updated, TRUE as perviously_engaged
+    FROM participants p
+    INNER JOIN participants_status ps ON
+      p.id = ps.participant_id
+      AND ps.status IN ('rejected', 'reject_ack')
+      AND ps.current = 'true'
+      AND ps.created_at < (NOW() - interval '6 month')
+    WHERE p.body->>'interested' = 'yes'
+    -- If anyone has a status in the past 6 months, exclude them
+    AND p.id NOT IN (
+      SELECT DISTINCT ON (ps.participant_id)
+      ps.participant_id
+      FROM participants_status ps
+      WHERE ps.created_at > (NOW() - interval '6 month')
+      AND ps.current = true
+      GROUP BY ps.participant_id
+    )
+    GROUP BY p.id;
+  `;
+  await dbClient.db.query(pushRejectedParticipants);
+};
+
+/**
+ * Gets all participants from temporary table who haven't been updated in 6 months
+ * Relies on createOpenParticipantsTable
+ * @returns {void}
+ */
+const getExpiredParticipants = async () => {
+  const getQuery = `
+    SELECT * FROM expired_participants_table
+    WHERE last_updated < (NOW() - interval '6 month');
+  `;
+  return dbClient.db.query(getQuery);
+};
+
+/**
+ * Drops temporary table created
+ */
+const dropOpenParticipantsTable = async () => {
+  const dropQuery = `
+    DROP TABLE expired_participants_table;
+  `;
+  return dbClient.db.query(dropQuery);
+};
+
+/**
+ * For all the expired participants, invalidates their statuses and withdraws the participant
+ * @returns {void}
+ */
+const expiredParticipantsStatusInvalidation = async () => {
+  const updateStatement = `DO $$
+    DECLARE
+      participant_rec RECORD;
+      status_rec RECORD;
+      data_obj JSONB;
+      body_obj JSONB;
+    BEGIN
+    FOR participant_rec IN
+      SELECT * FROM participants WHERE id IN (
+        SELECT id from expired_participants_table
+        WHERE last_updated < (NOW() - interval '6 month')
+      )
+    LOOP
+      -- Invalidates all current statuses for an expired participant
+      FOR status_rec IN SELECT * FROM participants_status WHERE participant_id = participant_rec.id AND current = true
+      LOOP
+        IF status_rec.data IS NULL THEN
+          data_obj = JSONB_SET('{}'::JSONB, '{cleanupDate}', to_jsonb(NOW()));
+        ELSE
+          data_obj = JSONB_SET(status_rec.data::JSONB, '{cleanupDate}', to_jsonb(NOW()));
+        END IF;
+        UPDATE participants_status SET data = data_obj, current = false WHERE id = status_rec.id;
+      END LOOP;
+      -- Withdraws participant and sets an expiration date
+      body_obj = JSONB_SET(participant_rec.body::JSONB, '{expiredDate}', to_jsonb(NOW()));
+      body_obj = body_obj || '{"interested" : "withdrawn"}';
+      UPDATE participants SET body = body_obj, updated_at = NOW() WHERE id = participant_rec.id;
```

Comment:
this looks like a bit of code duplication, there might not be a better way to do it perf-wise but couldn't we create a more generic function that invalidates statuses for a given list of users?

Reply:
What would that function look like? A piece of sql that we call in here? Or a separate query that we call after all this?

Reply:
I was thinking a loop that took in an array of IDs and then invalidated all participant statuses for those IDs

Also I didn't read this super closely but is it necessary to be a loop? 

why not
```
update participants_status set current = false where participant_id = foo
```

Reply:
Because in many cases, we're adding to the json data of a participant status to add the clean up date. I don't think that can be done in a general query for multiple users.